### PR TITLE
Allow NeuralNetClassifier to fit tensor labels #802 

### PR DESCRIPTION
**Pull request by TheAutumnOfRice**
Issue #802 
The original check_data method can only receive ndarray y-values because of np.unique function. Now the method is modified so that tensor y-values can be processed with torch.unique specifically.

### CODE REVIEW

File: skorch/classifier.py [hunk 1]
```diff
@@ -117,7 +117,14 @@ def check_data(self, X, y):
             raise ValueError(msg)
         if y is not None:
             # pylint: disable=attribute-defined-outside-init
-            self.classes_inferred_ = np.unique(y)
+            if isinstance(y,np.ndarray):
```
Comment (BenjaminBossan) [inline skorch/classifier.py:120]:
```suggestion
            if isinstance(y, np.ndarray):
```

File: skorch/classifier.py [hunk 2]
```diff
@@ -117,7 +117,14 @@ def check_data(self, X, y):
...
+                self.classes_inferred_ = np.unique(y)
+            elif isinstance(y,torch.Tensor):
```
Comment (BenjaminBossan) [inline skorch/classifier.py:122]:
```suggestion
            elif isinstance(y, torch.Tensor):
```

File: skorch/classifier.py [hunk 3]
```diff
@@ -117,7 +117,14 @@ def check_data(self, X, y):
...
+                self.classes_inferred_ = torch.unique(y)
+            else:
+                msg = (f"If y-values are given, they must be either numpy ndarray or "
+                       f"torch tensors, but not {type(y)}.")
+                raise ValueError(msg)
```
Comment (BenjaminBossan) [inline skorch/classifier.py:127]:
```suggestion
                raise TypeError(msg)
```

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks a lot for taking this!

There are a few stylistic changes you could make and I believe that `TypeError` is more fitting here.

Another thing: For this kind of change, we should have unit tests. Here, it is a bit tricky because the current code, i.e. using `np.unique` on a torch tensor, seems to work. The test might thus require a bit of patching.

If you need help with implementing the tests, feel free to tell me and I'll help you along the way.

**Comment by BenjaminBossan** [comment]
I see now that the change you made introduces an error elsewhere:

```python
    def test_classes_data_torch_tensor(self, net_cls, module_cls, data):
        X = torch.as_tensor(data[0])
        y = torch.as_tensor(np.arange(len(X)) % 10)
    
        net = net_cls(module_cls, max_epochs=0).fit(X, y)
>       assert (net.classes_ == [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]).all()
E       AttributeError: 'bool' object has no attribute 'all'
```

This is because `classes_` used to always be a numpy array but now it can be a torch tensor. Please make sure to cast the torch tensor to a numpy array (using `skorch.utils.to_numpy`).

**Comment by TheAutumnOfRice** [comment]
@BenjaminBossan Thanks for your replying!

Since `skorch.utils.to_numpy` is so powerful, we just need to add `to_numpy` to the original code, and everything is done :p

When it comes to unit tests, It's my first time to be required to use them, and I have no ideas how to deal with them haha. I hope this one-line modification can pass all these tests without my efforts.

**Comment by BenjaminBossan** [comment]
> Since `skorch.utils.to_numpy` is so powerful, we just need to add `to_numpy` to the original code, and everything is done :p

That's great, so this is enough to solve your original issue in #802? If so, yeah I believe we don't need extra tests :)

**Comment by TheAutumnOfRice** [comment]
Yeah, that works pretty well :)

**Comment by BenjaminBossan** [comment]
Great, thanks for your work @TheAutumnOfRice

**Pull request merged**